### PR TITLE
Cleanup some ScriptRunner & Settings code

### PR DESCRIPTION
**Pull request by dwijnand**
Follow up on #8564:

* Dedup setting outputDirs.setSingleOutput from outdir
* Reuse ClassPathFactory

### CODE REVIEW

File: src/compiler/scala/tools/nsc/ScriptRunner.scala [hunk 1]
```diff
@@ -12,12 +12,10 @@
 
 package scala.tools.nsc
 
-import scala.reflect.io.{Directory, File, Path}
-import scala.tools.nsc.CloseableRegistry
-import scala.tools.nsc.classpath.DirectoryClassPath
-import scala.tools.nsc.classpath.ZipAndJarClassPathFactory
+import scala.reflect.io.{ AbstractFile, Directory, File, Path }
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/ScriptRunner.scala:15]: Your braces setting is on the wrong side of history.

File: src/compiler/scala/tools/nsc/ScriptRunner.scala [hunk 2]
```diff
@@ -106,14 +104,7 @@ abstract class AbstractScriptRunner(settings: GenericRunnerSettings) extends Scr
...
     def hasClassToRun(location: Path): Boolean = {
-      val cp =
-        if (Jar.isJarOrZip(location)) {
-          import scala.reflect.io.AbstractFile
-          val fil = AbstractFile.getFile(location)
-          val reg = new CloseableRegistry
-          ZipAndJarClassPathFactory.create(fil, settings, reg)
-        }
-        else DirectoryClassPath(location.jfile)
+      val cp = ClassPathFactory.newClassPath(AbstractFile.getDirectory(location), settings)
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/ScriptRunner.scala:107]: Thanks, I have only a vague memory of chasing down how to use these pieces.

File: src/compiler/scala/tools/nsc/settings/ScalaSettings.scala [hunk 3]
```diff
@@ -58,7 +58,7 @@ trait ScalaSettings extends StandardScalaSettings with Warnings { _: MutableSett
   // argfiles is only for the help message
   /*val argfiles = */ BooleanSetting("@<file>", "A text file containing compiler arguments (options and source files)")
   val classpath     = PathSetting   ("-classpath", "Specify where to find user class files.", defaultClasspath) withAbbreviation "-cp" withAbbreviation "--class-path"
-  val outdir        = OutputSetting (outputDirs, ".").withPostSetHook(s => try outputDirs.setSingleOutput(s.value) catch { case FatalError(msg) => errorFn(msg) })
+  val outdir        = OutputSetting (".").withPostSetHook(s => try outputDirs.setSingleOutput(s.value) catch { case FatalError(msg) => errorFn(msg) }).tap(_.postSetHook())
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/settings/ScalaSettings.scala:61]: Together at last!

**Review by som-snytt — APPROVED** [review]
Now I can get some sleep.

**Comment by som-snytt** [comment]
We have to ask @SethTisue if this can go in now as clean-up of recent PR.

**Review by SethTisue — APPROVED** [review]

**Pull request merged**